PR to review https://github.com/rapidsai/cuml/pull/4407

levsnv opened the pull request:
PR to review [Universal Treelite tree walk function for FIL #4407](https://github.com/rapidsai/cuml/pull/4407)

levsnv committed:
side-by-side

levsnv committed:
style

levsnv committed:
converted max_depth to lambda form

levsnv committed:
remove test code

levsnv committed:
changed explicit tree accumulator to implicit (captured); fixed an obvious bug

levsnv committed:
debugging lambda-based conversion to FIL

levsnv committed:
fixed a bug due to swap child nodes/default left

levsnv committed:
improved walk_tree ease of use (syntactic sugar), converted bit_pool_size(...) to use walk_tree

levsnv committed:
shortened code

levsnv committed:
compiles

levsnv committed:
rewrote tree_depth_hist in terms of walk_tree

levsnv committed:
debugging changes

levsnv committed:
improved comments

levsnv committed:
removed tree walk start configurability - no need for now

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+
+template <typename T,
+          typename L,
+          typename DescentAccumulator,
```
I'm not sure `DescentAccumulator` is a good name. This type is used for two purposes (if non-empty):

1. A parameter of this type is passed into the node function (both inner and leaf).
2. For inner nodes, the function returns a pair of values of this type, one for the left and right child.

I think `ParameterT` or `ValueT` would be better names.

levsnv replied:
Since CS called the maximum distance between a root and leaf node "depth", I assume one descends a decision tree while inferring it.
If not this, I'd like the name to be more descriptive.

Meaningful alternatives I can think of: `RootToLeafPathState`, `InferenceTraversalState` (generalizing node index), 
I like `TraversalState` less, since some algorithms traverse decision trees in the opposite or fully incongruent order.
Note that the accumulator doesn't need to be reversible in this walk function and it does have a direction, so `PathState` and `DirectedPathState` sounds misleading.

I also am against ambiguous, `InnerNodeState`, `LeafState`.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+          typename InnerFunc,
+          typename LeafFunc>
+inline void walk_tree(const tl::Tree<T, L>& tree,
+                      DescentAccumulator desc_acc,
```
Again, not sure `desc_acc` is a good name. `value` would be better.

levsnv replied:
changed to `state`

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+          typename DescentAccumulator,
+          typename InnerFunc,
+          typename LeafFunc>
+inline void walk_tree(const tl::Tree<T, L>& tree,
```
A good documentation comment would definitely help understanding the parameters. E.g. what is the purpose of `DescentAccumulator`, what are the parameters and return types of `InnerFunc` and `LeafFunc`?

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
-  std::stack<pair_t> stack;
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  using stackable = struct {
```
`struct stack_element { ... }` or just `struct element { ... }`.

`using element = std::pair_t<int, DescentAccumulator>;` would also be fine.

levsnv replied:
tried `invitation`, since it takes this to visit a node. Note that the lambda types are updated to self-document that these are visit lambdas. `element` sounds like any `std::` container type parameter, which might be better for an implementation of the stack template, the vector template, etc.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  using stackable = struct {
+    int id;
```
Does `id` refer to the treelite or FIL node index? A documentation comment could clarify it.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
-    const pair_t& pair = stack.top();
-    int node_id        = pair.first;
-    int depth          = pair.second;
+    stackable node = stack.top();
```
I don't think `node` is the best name; an element of the stack containing the treelite node index together with some data is clearly not a node. However, I don't have a better name to suggest, so this is up to you.

levsnv replied:
will just use the first letter of the type, see discussion above

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -147,10 +176,9 @@ inline std::vector<cat_feature_counters> cat_counter_vec(const tl::Tree<T, L>& t
-    }
-  }
+    },
+    [](...) {});
```
```suggestion
    [](int) {});
```

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -159,21 +187,16 @@ template <typename T, typename L>
-    }
-  }
+    },
+    [](...) {});
```
```suggestion
    [](int) {});
```

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+          typename InnerFunc,
+          typename LeafFunc>
+inline void walk_tree(const tl::Tree<T, L>& tree,
+                      DescentAccumulator desc_acc,
```
Also, in every use of this function, the initial value is either `0` or `empty{}`, i.e. the default value of its type. Consider dropping the parameter altogether, or providing it with the default value.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+inline void walk_tree(const tl::Tree<T, L>& tree,
+                      DescentAccumulator desc_acc,
+                      InnerFunc inner_func,
+                      LeafFunc leaf_func)
```
Consider having a default `leaf_func` that does nothing (and a matching default `LeafFunc` for it).

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,64 @@ int tree_root(const tl::Tree<T, L>& tree)
+      if constexpr (std::is_empty<DescentAccumulator>()) {
+        inner_func(node.id);
+      } else {
+        std::tie(left_acc, right_acc) = inner_func(node.id, node.desc_acc);
```
I think that it can be more readable of `inner_func` accepts `left_acc` and `right_acc` as parameters passed by reference (or by pointer), but I'm not sure here.

levsnv replied:
see new implementation - a bit hazier at call site, but simpler in definition.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -349,69 +366,55 @@ int tree2fil(std::vector<fil_node_t>& nodes,
-
+      nodes[root + fil_node_id] = cs.node;
+
+      return std::pair<int, int>(left + cs.swap_child_nodes, left + !cs.swap_child_nodes);
```
`left + (cs.swap_child_nodes ? 1 : 0)`, and similar with the other expression.

levsnv replied:
is the new version simple?

canonizer reviewed (changes requested):
I like the direction where this is going, but there are still some comments.

1. Consider reducing the number of parameters in `walk_tree()`. `desc_acc` always has the default value of its type, and in some cases, the leaf function is empty, so a version with fewer parameters would be useful. It's up to you whether this is achieved by using default values, or by having a separate wrapper function with the same name.
2. Consider passing `left_acc` and `right_acc` into `inner_func` by reference to make it more readable.

levsnv committed:
addressed review comment. Mainly fixed argument count for lambdas and used empty type; renamed parameters and types

levsnv commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -200,16 +213,14 @@ cat_sets_owner allocate_cat_sets_owner(const tl::ModelImpl<T, L>& model)
 
-void adjust_threshold(
-  float* pthreshold, int* tl_left, int* tl_right, bool* default_left, tl::Operator comparison_op)
+void adjust_threshold(float* pthreshold, bool* swap_child_nodes, tl::Operator comparison_op)
```
I can no longer swap indices since they're not passed by reference. It actually simplifies the code, since swapping and related adjustments are done only once.

canonizer replied:
This is fine, and the new code looks cleaner.

levsnv commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -349,69 +356,53 @@ int tree2fil(std::vector<fil_node_t>& nodes,
       // left is the offset of the left child node relative to the tree root
       // in the array of all nodes of the FIL sparse forest
-      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * cur + 1 : sparse_index - root;
+      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * fil_node_id + 1 : sparse_index;
```
sparse_index should just be relative to tree, not forest. renames `cur` to `fil_node_id` because otherwise I kept forgetting what it was.

canonizer replied:
This is fine.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
+  inner_func(int node_id, RootToLeafPathState state) should return a pair of new states, one for
+each child node. leaf_func(int, RootToLeafPathState) returns nothing.
+**/
+template <typename RootToLeafPathState,
```
Just `PathState` would be better, `TraversalState` is also fine.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
+template <typename RootToLeafPathState,
+          typename Threshold,
+          typename Leaf,
+          typename InnerNodeVisitFunc,
```
I think `InnerFunc` and `LeafFunc` were better names.

levsnv replied:
sure. what about `visit_inner` and `visit_leaf` instead of `inner_func` and `leaf_func`?

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
+  /// needed to visit a node
+  struct invitation {
```
Not sure `invitation` is a good name, unless it's a standard term for such functions (`visit()` is definitely a standard term). I think `element` would be better.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
-      depth++;
-      ASSERT(depth < DEPTH_LIMIT, "depth limit reached, might be a cycle in the tree");
+    while (!tree.IsLeaf(i.tl_node_id)) {
+      auto [left_state, right_state] = inner_func(i.tl_node_id, i.state);
```
This loop is certainly more concise, but at the price of having to return a pair of empty values if we don't care about the state.

I think there should be a separate wrapper (also named `walk_tree()`) for the case when the state is empty. It can then wrap `leaf_func` into something that returns a tuple of empty values and accepts an extra empty argument.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
+inline int max_depth(const tl::Tree<T, L>& tree)
+{
+  int tree_depth = 0;
+  walk_tree<int>(
```
Now the state type has to be specified explicitly. I guess this it how it has to be, since it cannot be deduced from the function type.

levsnv replied:
the alternative, matching to a certain kind of `std::function` and wrapping the call site lambdas like this `std::function([](...){...})` looks more complicated to me.
I also tried
```c++
template <typename Threshold,
          typename Leaf,
          typename InnerNodeVisitFunc,
          typename TraversalState = typename std::invoke_result<InnerNodeVisitFunc, int, void*>::type::value_type,
          typename LeafNodeVisitFunc = void(int, TraversalState)>
inline void walk_tree(
  const tl::Tree<Threshold, Leaf>& tree,
  InnerNodeVisitFunc inner_func,
  LeafNodeVisitFunc leaf_func = [](int, TraversalState) {})
{
```

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -122,35 +151,27 @@ template <typename T, typename L>
-        cat_feature_counters& counters = res[tree.SplitIndex(node_id)];
-        counters =
-          cat_feature_counters::combine(counters, cat_feature_counters{max_matching_cat, 1});
+  walk_tree<empty>(tree, [&](int node_id, empty val) {
```
`empty val` clearly looks redundant (though it's necessary for the current solution).

Consider providing a wrapper version of `walk_tree` which accepts functions without additional dummy arguments.

Alternatively, you could use the previous version. It's definitely better for the function to be more readable at the call site, the definition is going to be heavily templatized anyway.

levsnv replied:
that's the best design! done.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -122,35 +151,27 @@ template <typename T, typename L>
-      node_id = tree.RightChild(node_id);
+      cat_feature_counters& counters = res[tree.SplitIndex(node_id)];
+      counters = cat_feature_counters::combine(counters, cat_feature_counters{max_matching_cat, 1});
+      return std::pair<empty, empty>();
```
Again, consider providing a wrapper version of `walk_tree()` that doesn't expect the function to return any value.

Alternatively, you could use some parts from the previous version. It's definitely better for the function to be more readable at the call site, the definition is going to be heavily templatized anyway.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -159,21 +180,13 @@ template <typename T, typename L>
-      }
-      stack.push(tree.LeftChild(node_id));
-      node_id = tree.RightChild(node_id);
+  walk_tree<empty>(tree, [&](int node_id, empty val) {
```
Same here.

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -349,69 +356,53 @@ int tree2fil(std::vector<fil_node_t>& nodes,
-
+      nodes[root + fil_node_id] = cs.node;
+
+      return cs.swap_child_nodes ? std::tuple(left + 1, left) : std::tuple(left, left + 1);
```
This is definitely more readable.

Btw, won't the following work?

```suggestion
      return cs.swap_child_nodes ? {left + 1, left} : {left, left + 1};
```

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -69,32 +69,61 @@ int tree_root(const tl::Tree<T, L>& tree)
+      // e.g. cycles in the forest
+      const int DEPTH_LIMIT = 500;
+      ASSERT(node_depth < DEPTH_LIMIT, "node_depth limit reached, might be a cycle in the tree");
+      return std::tuple(node_depth + 1, node_depth + 1);
```
```suggestion
      return {node_depth + 1, node_depth + 1};
```

This should work, and is definitely more readable.

levsnv replied:
oooh, it's a list_initializer type now, isn't it? I'm always up for omitting types where a small list of things is all the humans expect!

canonizer commented on cpp/src/fil/treelite_import.cu:
```diff
@@ -349,69 +356,53 @@ int tree2fil(std::vector<fil_node_t>& nodes,
-    if (static_cast<std::size_t>(depth) >= hist.size()) hist.resize(depth + 1, {0, 0});
-    hist[depth].n_leaves++;
-  }
+      return std::tuple(depth + 1, depth + 1);
```
```suggestion
      return {depth + 1, depth + 1};
```

levsnv replied:
doesn't work. From what I've read, it shouldn't work in any return context, but I added this wherever the compiler let me sneak it in.
What is the type of such a thing anyway? namely, isn't there an issue with deducing a reference type initializer_list in a return context?

canonizer reviewed (approved):
Nice work! Approved provided that the comments are addressed.

Some items I'd like to highlight: 
1. Could you use `return {a, b};` when returning a tuple? This should work, and is less cluttered.
2. Would it be possible to provide a wrapper for the case of empty state? In this case, the  definition would be concise, and the call sites won't be cluttered with dummy arguments and return statements.

levsnv committed:
addressed other review comments

levsnv committed:
simplified

levsnv committed:
addressed remaining review comments

levsnv committed:
tried self-wrapping function

levsnv committed:
remove lazy

levsnv committed:
a bit more readable

levsnv committed:
Merge branch 'branch-22.02' of github.com:rapidsai/cuml into tree-lambda

pentschev committed:
Update to UCX-Py 0.24 (#4396)



Authors:
  - Peter Andreas Entschev (https://github.com/pentschev)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4396

achirkin committed:
Add a warning to prefer LinearSVM over SVM(kernel='linear') (#4382)

Suggest using LinearSVM when the user chooses to use the linear kernel in SVM. The reason is that LinearSVM uses a specialized faster solver.

Closes #1664
Also partially addresses #2857

Authors:
  - Artem M. Chirkin (https://github.com/achirkin)

Approvers:
  - Tamas Bela Feher (https://github.com/tfeher)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4382

cjnolet committed:
Hiding cusparse deprecation warnings (#4373)



Authors:
  - Corey J. Nolet (https://github.com/cjnolet)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4373

Commit:
fix minor ASAN issues in UMAPAlgo::Optimize::find_params_ab() (#4405)

There were actuall 2 minor issues that prevented `UMAPAlgo::Optimize::find_params_ab()` from being ASAN-clean at the moment:

- One is the mem leaks, of course
- Another one is the `malloc()`-`delete` mismatch -- only memory allocated using `new` or equivalent should be freed with operator `delete` or `delete[]`

Another issue that was also addressed here: exception safety (i.e., by using `make_unique` from C++-14)

Signed-off-by: Yitao Li <yitao@rstudio.com>

Authors:
  - Yitao Li (https://github.com/yitao-li)

Approvers:
  - Zach Bjornson (https://github.com/zbjornson)
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4405

zbjornson committed:
Simplify perplexity normalization in t-SNE (#4425)

P_sum is equal to n. See #2622 where I made this change once before. #4208 changed it back while consolidating code.

Authors:
  - Zach Bjornson (https://github.com/zbjornson)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4425

venkywonka committed:
RF: code re-organization to enhance build parallelism (#4299)

This PR separates the Decision tree kernels into separate Translation Units (TU) and explicitly instantiates templates. 
This is helpful in 2 ways:
1. refactoring top-level RF/DT code now would not require recompilation of the kernels 
2. Since they are separated into different TUs and linked, they can leverage build parallelism (4x improvement in rebuild times after touching kernel definitions)

Rebuilding by running `time ./build.sh libcuml -v -n PARALLEL_LEVEL=20` after touching RF kernels comparison:
(Note: using `--ccache` doesn't matter here, assuming after touching RF kernels the state of the code-base is completely new and not part of ccache's hashed index)
<details><summary>This PR</summary>
  
  ```
  real    0m20.054s             
  user    2m28.436s                                               
  sys     0m14.241s 
  ```
</details>
<details><summary>branch-21.12</summary>
  
  ```
real    1m21.197s                                                                                                                    
user    2m5.751s                                                                                                                     
sys     0m6.050s
  ```
</details>

Some other changes include renaming and reorganizing files, pruning headers and cleaning up some code

Things to do: 
- [x] split DT Kernels
- [x] benchmark for regressions

Authors:
  - Venkat (https://github.com/venkywonka)

Approvers:
  - Rory Mitchell (https://github.com/RAMitchell)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4299

viclafargue committed:
Remove comment numerical warning (#4408)

Answers #4203
Just set in stone the warning filter for "Numerical issues".

Authors:
  - Victor Lafargue (https://github.com/viclafargue)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4408

hcho3 committed:
Fix docstring for npermutations in PermutationExplainer (#4402)

Closes #4047

Authors:
  - Philip Hyunsu Cho (https://github.com/hcho3)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4402

cjnolet committed:
Using sparse public API functions from RAFT (#4389)



Authors:
  - Corey J. Nolet (https://github.com/cjnolet)

Approvers:
  - Divye Gala (https://github.com/divyegala)
  - Dante Gama Dessavre (https://github.com/dantegd)

URL: https://github.com/rapidsai/cuml/pull/4389

dantegd committed:
Remove direct Scikit-learn imports (#4431)

* FIX Remove hard sklearn imports

* FIX Missing whitespace

* FIX minor error

* FIX PEP8 fixes

GPUtester committed:
Merge pull request #4434 from rapidsai/branch-21.12

[gpuCI] Forward-merge branch-21.12 to branch-22.02 [skip gpuci]

divyegala committed:
Integrating RAFT handle updates (#4313)



Authors:
  - Divye Gala (https://github.com/divyegala)
  - Corey J. Nolet (https://github.com/cjnolet)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4313

jjacobelli committed:
Update ucx-py version on release using rvc (#4411)

Update `ucx-py` version on release using `rvc`

Authors:
  - Jordan Jacobelli (https://github.com/Ethyling)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)

URL: https://github.com/rapidsai/cuml/pull/4411

bdice committed:
Update CUDA 11.5 conda environment to use 22.02 pinnings. (#4450)

This PR updates the pinnings of the conda environment for CUDA 11.5 to use 22.02 packages. This resolves conflicts between a5e7cfb5c32f6e7eb7e8d64628f686c3b973789e and #4364.

Authors:
  - Bradley Dice (https://github.com/bdice)

Approvers:
  - Dante Gama Dessavre (https://github.com/dantegd)
  - AJ Schmidt (https://github.com/ajschmidt8)

URL: https://github.com/rapidsai/cuml/pull/4450

cjnolet committed:
Use RAFT template instantations for distances (#4302)



Authors:
  - Corey J. Nolet (https://github.com/cjnolet)
  - Dante Gama Dessavre (https://github.com/dantegd)

Approvers:
  - Divye Gala (https://github.com/divyegala)
  - Dante Gama Dessavre (https://github.com/dantegd)
  - Jordan Jacobelli (https://github.com/Ethyling)

URL: https://github.com/rapidsai/cuml/pull/4302

levsnv committed:
Merge branch 'branch-22.02' of github.com:rapidsai/cuml into tree-lambda

jjacobelli committed:
Add missing imports tests (#4452)

Add missing imports tests

Authors:
  - Jordan Jacobelli (https://github.com/Ethyling)

Approvers:
  - AJ Schmidt (https://github.com/ajschmidt8)

URL: https://github.com/rapidsai/cuml/pull/4452

levsnv committed:
Merge branch 'unify-dense-sparse-import' into tree-lambda

achirkin committed:
Move NVTX range helpers to raft (#4445)

Move NVTX range helpers to raft  and extend them a little bit.
Corresponding raft PR: https://github.com/rapidsai/raft/pull/416 .

Authors:
  - Artem M. Chirkin (https://github.com/achirkin)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuml/pull/4445

levsnv committed:
Unify dense and sparse import in FIL (#4328)



Authors:
  - Levs Dolgovs (https://github.com/levsnv)

Approvers:
  - Andy Adinets (https://github.com/canonizer)
  - William Hicks (https://github.com/wphicks)

URL: https://github.com/rapidsai/cuml/pull/4328

levsnv committed:
Merge branch 'branch-22.02' into tree-lambda

levsnv committed:
Apply suggestions from code review

Co-authored-by: William Hicks <wphicks@users.noreply.github.com>
